src/main/scala/com/databricks/labs/overwatch/pipeline/PostProcessor.scala

```diff
@@ -1,28 +1,145 @@
+   * Depending on the scopes several of the targets may not exist. The scopes could be disabled and / or the module
+   * may have failed since first run resulting in no table / etc.
+   * @return All pipeline targets marked for optimize less those that do not exist.
+   */
+  private def validateExists: Array[PipelineTable] = {
+    val targetsPar = tablesToOptimize.toArray.par
+    val taskSupport = new ForkJoinTaskSupport(new ForkJoinPool(16))
+    targetsPar.tasksupport = taskSupport
+    targetsPar.filter(_.exists).toArray
+  }
+
+  /**
+   * launch the optimize jobs in parallel
+   */
+  private def executeOptimize: Unit = {
+    appendIntermediateTargets
+    val validatedTargets = validateExists
+    val invalidTargets = tablesToOptimize.map(_.tableLocation).diff(validatedTargets.map(_.tableLocation))
+    val invalidTargetsMsg = s"OPTIMIZE: Missing Targets: \n${invalidTargets.mkString("\n")}"
+    val msg = s"OPTIMIZE: Targets To Optimize: \n${validatedTargets.map(_.tableLocation).mkString("\n")}"
+
+    if (config.debugFlag) {
+      println(invalidTargetsMsg)
+      println(msg)
+    }
+    logger.log(Level.INFO, invalidTargetsMsg)
+    logger.log(Level.INFO, msg)
     // TODO -- spark_events_bronze -- put in proper rules -- hot fix due to optimization issues
-    Helpers.parOptimize(tablesToOptimize.toArray.filterNot(_.name == "spark_events_bronze"), maxFileSizeMB = 128)
-    Helpers.parOptimize(tablesToOptimize.toArray.filter(_.name == "spark_events_bronze"), maxFileSizeMB = 32)
+    Helpers.parOptimize(validatedTargets.filterNot(_.name == "spark_events_bronze"), maxFileSizeMB = 128)
+    Helpers.parOptimize(validatedTargets.filter(_.name == "spark_events_bronze"), maxFileSizeMB = 32)
+  }
+
+  /**
+   * When the optimize is not executed from within the pipeline run structure it's necessary to go back and
+   * update the latest "lastOptimizedTS" state of each module to reflect this optimize as the latest optimization complete
+   */
+  private def updateLastOptimizedDate: Unit = {
+    val lastOptimizedUpdateDF = Optimizer.getLatestSuccessState(config.databaseName)
+      .withColumn("lastOptimizedTS", lit(System.currentTimeMillis()))
+      .alias("updates")
+
+    val immutableCols = (pipelineStateTarget.keys ++ pipelineStateTarget.partitionBy ++ pipelineStateTarget.incrementalColumns).distinct
+    val mergeCondition = immutableCols.map(k => s"updates.$k = target.$k").mkString(" AND ")
+    val updateExpr = Map("target.lastOptimizedTS" -> "updates.lastOptimizedTS")
+
+    val msg =
+      s"""
+         |Pipeline State Update -- Updating Last Optimization Timestamp
+         |MERGE CONDITION: $mergeCondition
+         |UPDATE EXPRESSION: $updateExpr
+         |""".stripMargin
+
+    if (config.debugFlag) println(msg)
+    logger.log(Level.INFO, msg)
+    DeltaTable.forPath(pipelineStateTarget.tableLocation).alias("target")
+      .merge(lastOptimizedUpdateDF, mergeCondition)
+      .whenMatched
+      .updateExpr(updateExpr)
```

Comment:
what will happen when we have new tables added into list to optimize?

Reply:
If they are targets that result from a module, there will be another record in the pipReport. If future tables are supporting tables they will be added to the optimize of the parent target they support. Either way this merge should continue to work as expected as far as I can tell, did I miss something?